core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java

```diff
@@ -58,17 +58,21 @@
     private final ConcurrentMap<Class, BeanInfo> beanInfoCache = new ConcurrentHashMap<>();
     private TypeConverter defaultConverter;
 
-    private boolean devMode = false;
+    private boolean devMode;                       // Defaults to false
+    private boolean devModeValueSet;               // Defaults to false (not explicitly set yet)
     private boolean enableExpressionCache = true;
-    private boolean enableEvalExpression;
+    private boolean enableEvalExpression;          // Defaults to false
+    private boolean enableEvalExpressionValueSet;  // Defaults to false (not explicitly set yet)
```

Comment:
Sorry you did a lot of work but please replace all change-limiters logic including their unit test with proposed solution at [here](https://github.com/apache/struts/pull/292#discussion_r241949472).